core/pkg/ingress/controller/backend_ssl.go

```diff
@@ -108,19 +110,24 @@ func (ic *GenericController) getPemCertificate(secretName string) (*ingress.SSLC
 	}
 
 	secret := secretInterface.(*api.Secret)
-	cert, ok := secret.Data[api.TLSCertKey]
-	if !ok {
-		return nil, fmt.Errorf("secret named %v has no private key", secretName)
-	}
-	key, ok := secret.Data[api.TLSPrivateKeyKey]
-	if !ok {
-		return nil, fmt.Errorf("secret named %v has no cert", secretName)
-	}
+	cert, okcert := secret.Data[api.TLSCertKey]
+	key, okkey := secret.Data[api.TLSPrivateKeyKey]
 
 	ca := secret.Data["ca.crt"]
 
 	nsSecName := strings.Replace(secretName, "/", "-", -1)
-	s, err := ssl.AddOrUpdateCertAndKey(nsSecName, cert, key, ca)
+
+	var s *ingress.SSLCert
+	if okcert && okkey {
+		glog.V(3).Infof("Found certificate and private key, configuring %v as a TLS Secret", secretName)
+		s, err = ssl.AddOrUpdateCertAndKey(nsSecName, cert, key, ca)
```

Comment:
Is it possible to have a single function replace AddOrUpdateCertAndKey/AddOrUpdateCertAuth, called AddTLSKeyPair(nsSecName, cert, key, ca), which if cert and key are non-empty, appends ca to cert like AddOrUpdateCertAndKey and if they're empty, writes ca to file directly like AddOrUpdateCertAuth ? (I haven't though too much about the refactor so feel free to tell me it's too hard, just seems confusing to have 2 similar functions)